src/number_theory/kummer_dedekind.lean

```diff
@@ -0,0 +1,328 @@
+then `R[α] / (I) = (R[x] / (f)) / pS = (R/p)[x] / (f mod p)` -/
+noncomputable def power_basis.quotient_equiv_quotient_minpoly_map [is_domain R] [is_domain S]
+  (pb : power_basis R S) (I : ideal R)  :
+  (S ⧸ I.map (algebra_map R S)) ≃ₐ[R] (polynomial (R ⧸ I)) ⧸
+    (ideal.span ({(minpoly R pb.gen).map I^.quotient.mk} : set (polynomial (R ⧸ I)))) :=
+alg_equiv.trans
+  (alg_equiv.of_ring_equiv
+    (ideal.quotient_equiv _ (ideal.map (adjoin_root.of (minpoly R pb.gen)) I)
+    (adjoin_root.equiv' (minpoly R pb.gen) pb
+    (by rw [adjoin_root.aeval_eq, adjoin_root.mk_self])
+    (minpoly.aeval _ _)).symm.to_ring_equiv
+    (by rw [ideal.map_map, alg_equiv.to_ring_equiv_eq_coe, ← alg_equiv.coe_ring_hom_commutes,
+            ← adjoin_root.algebra_map_eq, alg_hom.comp_algebra_map]))
+  (λ x, by rw [← ideal.quotient.mk_algebra_map, ideal.quotient_equiv_apply,
+    ring_hom.to_fun_eq_coe, ideal.quotient_map_mk, alg_equiv.to_ring_equiv_eq_coe,
+    ring_equiv.coe_to_ring_hom, alg_equiv.coe_ring_equiv, alg_equiv.commutes,
+    quotient.mk_algebra_map]))
+  (adjoin_root.quot_equiv_quot_map _ _)
+
+@[simp] lemma power_basis.quotient_equiv_quotient_minpoly_map_apply [is_domain R] [is_domain S]
+  (pb : power_basis R S) (I : ideal R) (x : polynomial R) :
+  pb.quotient_equiv_quotient_minpoly_map I (ideal.quotient.mk _ (aeval pb.gen x)) =
+    ideal.quotient.mk _ (x.map I^.quotient.mk) :=
+by rw [power_basis.quotient_equiv_quotient_minpoly_map, alg_equiv.trans_apply,
+    alg_equiv.of_ring_equiv_apply, quotient_equiv_mk, alg_equiv.coe_ring_equiv',
+    adjoin_root.equiv'_symm_apply, power_basis.lift_aeval,
+    adjoin_root.aeval_eq, adjoin_root.quot_equiv_quot_map_apply]
+
+variable [decidable_eq (ideal S)]
+
+noncomputable instance {I: ideal R} [hI : is_maximal I] : field (R ⧸ I) :=
+((ideal.quotient.maximal_ideal_iff_is_field_quotient I).mp hI).to_field
+
+namespace kummer_dedekind
+
+variables [is_domain R] [is_dedekind_domain R] [is_domain S] [is_dedekind_domain S] [algebra R S]
+variables  (pb : power_basis R S) {I : ideal R}
+
+noncomputable def factors_equiv (hI : is_maximal I) :
+  {J : ideal S | J ∣ I.map (algebra_map R S) } ≃o
+    {J : ideal (polynomial $ R ⧸ I ) | J ∣ ideal.span { map I^.quotient.mk (minpoly R pb.gen) }} :=
+ideal_factors_equiv_of_quot_equiv (pb.quotient_equiv_quotient_minpoly_map I)
+
+lemma factors_equiv_is_dvd_iso (hI : is_maximal I) (l l' : ideal S)
+  (hl : l ∣  I.map (algebra_map R S)) (hl' : l'∣ I.map (algebra_map R S) ) :
+  (factors_equiv pb hI ⟨l, hl⟩ : ideal (polynomial $ R ⧸ I )) ∣
+    (factors_equiv pb hI ⟨l', hl'⟩) ↔ l ∣ l' :=
+begin
+  suffices : factors_equiv pb hI ⟨l', hl'⟩ ≤ factors_equiv pb hI ⟨l, hl⟩
+    ↔ (⟨l', hl'⟩ : {J : ideal S | J ∣ I.map (algebra_map R S)}) ≤ ⟨l, hl⟩,
+  rwa [subtype.mk_le_mk, ← subtype.coe_le_coe, ← dvd_iff_le, ← dvd_iff_le] at this,
+  exact (factors_equiv pb hI).le_iff_le,
+end
+
+lemma mem_normalized_factors_factors_equiv_of_mem_normalized_factors (hI : is_maximal I)
```

Comment:
These lemmas are a little hard to read. Do you mind adding a docstring if they're mathematical meaningful?